testng-core/src/test/java/test/listeners/issue2916/LogContainer.java

```diff
@@ -0,0 +1,34 @@
+package test.listeners.issue2916;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+import org.testng.internal.AutoCloseableLock;
+
+public enum LogContainer {
+  instance;
```

Comment:
single instance per JVM would still have issues with concurrent test execution. For instance, if multiple tests execute at the same time, they would log into the same `LogContainer` thus they would corrupt each others' logs

Reply:
True that. Some additional context:

* The tests are not running in parallel.
* The tests in question basically work in 2 modes [ API mode wherein we feed in the listener instance and CLI mode wherein the listener instances are created by TestNG using reflection ]
* When working with CLI mode, I am not sure as to how to access the listener instance that TestNG created so that I can access the logs captured by the test.
* I seem to be seeing a pattern in the failure. The failure is always confined to the tests with timeouts. When we run the test with the StressLCM/StressGCM flags, the tests don't seem to be timing out, but instead they are running to completion as seen by the below output

```

but some elements were not found: |  
-- | --
  | ["DragonWarrior.onTestFailedWithTimeout", |  
  | "MasterShifu.onTestFailedWithTimeout", |  
  | "MasterOogway.onTestFailedWithTimeout"] |  
  | and others were not expected: |  
  | ["DragonWarrior.onTestSuccess", |  
  | "MasterShifu.onTestSuccess", |  
  | "MasterOogway.onTestSuccess"]

```

This makes me think that there's something wrong with the codebase when it comes to running timing out tests, but am not able to figure out what, since I dont know what to look for. I am looking for static states (non existent). The shared states are being tracked via threadsafe variants (Atomic*).

Any other pointers you can suggest ?

Reply:
[This output](https://scans.gradle.com/s/jqd7f7ndmlksy/tests/task/:testng-core:test/details/test.listeners.ListenersTest/ensureOrderingForTestListenersViaXmlTag?top-execution=2) confirms my hunch. When running with the stress flags, a test that is supposed to be timing out, does not time out. Not sure what will cause that because the code paths that are related to timeouts dont have any global variables etc.,

Reply:
@juherr @vlsi - I have hit a dead end with this. I need some help so that I can proceed further.

Reply:
> When running with the stress flags, a test that is supposed to be timing out, does not time out

What is the test that should time out and fails to do so?
Have you investigated if the timeout logic is implemented properly? It might be there is a concurrency issue within the timeout implementation.

Reply:
@vlsi - Its a very simple test that uses the `timeout` feature that TestNG provides.

[This](https://github.com/testng-team/testng/blob/e5e22a715c61e11574d0fa3931d98b5b31b80f6e/testng-core/src/test/java/test/listeners/issue2916/ElaborateSampleTestCase.java#L27-L33) is what the test looks like

```java
  @Test(timeOut = 25, priority = 4)
  public void timingOutTest() throws InterruptedException {
    TimeUnit.MILLISECONDS.sleep(100);
    new Throwable().printStackTrace();
    if (counter != 3) {
      Assert.fail();
    }
  }
```

The test passed consistently when I don't use those flags. That is what makes me believe that when those flags are enabled, TestNG's approach of basically running these tests has a problem. But no matter how many times I have skimmed through them, I still am not able to identify them :( 

[This](https://github.com/testng-team/testng/blob/e5e22a715c61e11574d0fa3931d98b5b31b80f6e/testng-core/src/main/java/org/testng/internal/invokers/MethodInvocationHelper.java#L295-L313) is where TestNG is basically running those tests driven/controlled by timeouts.

Reply:
Well, having timeout of `25ms` vs `sleep` of `100ms` might be too optimistic. In other words, "sleep 100ms" within the test might be too small amount of time to "guarantee" the interrupt logic would have enough time to execute.

Could you try increasing `sleep` to 10sec?

Reply:
@vlsi - Lol.. I should have started off with this as the question. It could have atleast saved me a day's time. Bumping up the sleep time worked. I am now going to create a new PR from `master` again, with just the timeout bumped to see if the fix is consistent. If it is, then I think I will perhaps want to stick with just that change instead of this elaborate change which still doesn't guarantee thread safety when the tests are run in parallel (which you also rightly called out). 

Because there's reflection and since there's no straightforward way of retrieving all the logs from the listeners (especially when the test invokes TestNG via the CLI), I will want the changeset to be as little as possible.

Reply:
Thank you so much @vlsi for spending time and helping me out.